ext/standaloneusers/Civi/Api4/Action/User/WriteTrait.php

```diff
@@ -0,0 +1,103 @@
+   */
+  protected $hashPassword = 'safe';
+
+  /**
+   * If given, this is checked against the current user's password before the change is allowed.
+   *
+   * @param string
+   */
+  protected $actorPassword;
+
+  protected function validateValues() {
+    if ($this->getCheckPermissions()) {
+      // We must have a logged in user...
+      $loggedInUserID = \CRM_Utils_System::getLoggedInUfID() ?? FALSE;
+      if (!$loggedInUserID) {
+        // Something weird going on. This codepath should never happen.
+        throw new UnauthorizedException("Not logged on");
+      }
+
+      // If changing someone else's password, require 'cms:administer users'
+      $values = $this->getValues();
+      if (($values['id'] ?? NULL) !== $loggedInUserID
+        && !\CRM_Core_Permission::check(['cms:administer users'])) {
+        throw new UnauthorizedException("You are not permitted to change users' passwords.");
+      }
+
+      // Check that we have the logged-in-user's password.
+      $security = \Civi\Standalone\Security::singleton();
+      $isAuthorized = FALSE;
+      if ($this->actorPassword) {
+        $storedHashedPassword = \Civi\Api4\User::get(FALSE)
+        ->addWhere('id', '=', $loggedInUserID)
+        ->addSelect('password')
+        ->execute()
+        ->first()['password'];
+        if (!$security->checkPassword($this->actorPassword, $storedHashedPassword)) {
+          throw new UnauthorizedException("Incorrect password");
+        }
+      }
+      // @todo contact checksum/JWT for password resets.
+      // elseif ($this->jwt) { ... }
+      else {
+        throw new UnauthorizedException("Unauthorized");
+      }
+    }
+    parent::validateValues();
+  }
+
+  /**
+   *
+   * @param array $record
+   * @throws \CRM_Core_Exception
+   */
+  protected function formatWriteValues(&$record) {
+    if (isset($record['password'])) {
+      if ($this->hashPassword === 'hash_password'
+        || ($this->hashPassword === 'safe' && $this->passwordLooksLikePlaintext($record['password']))
+        ) {
+        $security = \Civi\Standalone\Security::singleton();
+        $record['password'] = $security->hashPassword($record['password']);
+      }
+    }
+    parent::formatWriteValues($record);
+  }
+
+  /**
+   * Heuristic to determine if a password looks unhashed.
+   *
+   * This should not be relied upon because it prevents people (successfully) using
+   * passwords that match this pattern!
+   *
+   * It's here just in case someone does a User::get and then a User::save
+   * without thinking about it; in this case it would prevent rehashing (and thereby
+   * invalidating) the loaded already-hashed passwords.
+   *
+   * Drupal7, WordPress at least both have passwords like $<something>$...
+   * As we only support Drupal7 style passwords at the mo, this will do,
+   * but may need updating later.
+   */
+  protected function passwordLooksLikePlaintext(string $pass): bool {
+    return !preg_match('/^[$].{1,10}[$].+/', $pass);
```

Comment:
Right, matching the prefix `$` means that the heuristic can be wrong if a user password organically starts with `$`. Hence the good advice to be clear about the request's hashing expectation.

If the edge feels itchy and bothersome, then these might soothe them:

* Generally prohibit `$` in the first character of a password. (Include that as part of password-validation.)
* Change to a different signifier.  (Ex: `CryptoToken` uses ([`\u0002CTK?`](https://github.com/civicrm/civicrm-core/blob/5.66/Civi/Crypto/CryptoToken.php#L33-L42), which is non-conflicted for plain-text fields. It's somewhat different data/use-case, but a prefix like `\u0002PHC$` would remove ambiguity. Though data-migrations would need to inject the extra signifier, and I think it's nice that the current on-disk format can be copied without any filters)